src/repo-service.cpp

```diff
@@ -138,3 +295,132 @@ void RepoService::openLocalFile(const QString& repo_id,
         helper->start();
     }
 }
+
+void RepoService::onGetRequestSuccess(const ServerRepo& repo)
+{
+    GetRepoRequest* req = qobject_cast<GetRepoRequest*>(sender());
+    if (!req)
+        return;
+
+    SyncedSubfolderhasRepoID subfolder_helper(repo.id);
+    std::vector<SyncedSubfolder>::iterator pos = std::find_if(synced_subfolders_.begin(), synced_subfolders_.end(), subfolder_helper);
+
+    if (pos != synced_subfolders_.end()) {
+        ServerRepo fixed_repo = repo;
+        fixed_repo.parent_path = pos->parent_path();
+        fixed_repo.parent_repo_id = pos->parent_repo_id();
+        server_repos_.push_back(fixed_repo);
+    }
+
+    delete req;
+    get_repo_reqs_.pop_front();
+
+    if (!get_repo_reqs_.empty())
+        get_repo_reqs_.front()->send();
+    else
+        emit refreshSuccess(server_repos_);
+}
+
+void RepoService::onGetRequestFailed(const ApiError& /*error*/)
+{
+    GetRepoRequest* req = qobject_cast<GetRepoRequest*>(sender());
+    if (!req)
+        return;
+
+    delete req;
+    get_repo_reqs_.pop_front();
+
+    if (!get_repo_reqs_.empty())
+        get_repo_reqs_.front()->send();
+    else
+        emit refreshSuccess(server_repos_);
+}
+
+void RepoService::startGetRequestFor(const QString &repo_id)
+{
+    bool was_empty = get_repo_reqs_.empty();
+    GetRepoRequest *req = new GetRepoRequest(seafApplet->accountManager()->currentAccount(), repo_id);
+    get_repo_reqs_.push_back(req);
+    connect(req, SIGNAL(success(const ServerRepo&)), this, SLOT(onGetRequestSuccess(const ServerRepo&)));
+    connect(req, SIGNAL(failed(const ApiError&)), this, SLOT(onGetRequestFailed(const ApiError&)));
+    if (was_empty)
```

Comment:
`get_repo_reqs_` is a `std::list` aka double-link list. If calling `std::list::length` is avoidable,  using it is not so smart a option that it will iterates all its notes.